### ISSUE TITLE
[BUG]: `Dict._find_index` doesn't check all index slots, can infinite loop

### ISSUE THREAD

**Issue by mzaks** [modularml/mojo#1729]
### Bug description

While executing the code listed in reproduce section I experience a crash with following output:
```
Stack dump:
0.      Program arguments: /home/maxim/.modular/pkg/packages.modular.com_mojo/bin/mojo /home/maxim/dev/mojo_hash/test_hashmap.mojo
#0 0x000055abe8b7b937 (/home/maxim/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x622937)
#1 0x000055abe8b7950e (/home/maxim/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x62050e)
#2 0x000055abe8b7c00f (/home/maxim/.modular/pkg/packages.modular.com_mojo/bin/mojo+0x62300f)
#3 0x00007f9a2cc42520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
#4 0x00007f99b0003bdd 
Segmentation fault (core dumped)
```

### Steps to reproduce

```
    let keys = String('Проснувшись однажды утром после беспокойного сна, Грегор Замза').split(" ")
    var d = Dict[StringKey, Int]()
    for i in range(len(keys)):
        d[keys[i]] = i
```

### System information

```shell
Ubuntu 22.04.3 LTS
mojo 0.7.0 (af002202)
modular 0.4.1 (2d8afe15)
```

**Comment by JoeLoser** [comment]
Closing as this won't work until we address https://github.com/modularml/mojo/issues/1616.  Please feel free to re-open if you disagree and we can discuss.

**Comment by JoeLoser** [comment]
Oops, you're right — sorry about that.  This has already been fixed internally.  Both of your repros work as expected now.  Here's a modified example that works internally (spoiler: `String` has a hash function and so on now to avoid `StringKey`).

```
 cat test.mojo
from collections import Dict, KeyElement

def main():
    # let keys = String('Проснувшись однажды утром после беспокойного сна, Грегор Замза').split(" ")
    let keys = String('Проснувшись однажды утром после беспокойного сна Грегор Замза').split(" ") # this works too
    var d = Dict[String, Int]()
    for i in range(len(keys)):
        d[keys[i]] = i
    print(d.size, len(keys))
    for i in range(len(keys)):
        let k = keys[i]
        print(k, d[k])
```

This prints:

```
8 8
Проснувшись 0
однажды 1
утром 2
после 3
беспокойного 4
сна, 5
Грегор 6
Замза 7
```

So, going to keep this as closed and it'll be available in the next release soon.

**Comment by mzaks** [comment]
@JoeLoser the bug is back in Mojo 24.1.1

**Comment by JoeLoser** [comment]
> @JoeLoser the bug is back in Mojo 24.1.1, but only reproduce able on Intel i7. On Apple M1, everything works fine.

Thanks for letting us know.  Do you mind opening a new GH issue and mentioning it's a regression (only applies on Intel, etc.) and we'll take a look?  You'll also be able to see the `dict` code yourself soon and maybe even fix it too if you're up for it 😄

**Comment by mzaks** [comment]
Hi @JoeLoser I figured out the issue. It’s because of the random probing implemented here https://github.com/modularml/mojo/blob/457378c3c5bafa287d3fc6be3a913e71e2233231/stdlib/src/collections/dict.mojo#L678

For the test case I provided the probing needs 11 rounds to find an empty slot, where there are 8 reserved slots and we add sixth item. 
As far as I understand the Python algorithm looks for a slot in a while True: loop, where Mojo reimplement loops for the number of reserved slots. https://github.com/modularml/mojo/blob/457378c3c5bafa287d3fc6be3a913e71e2233231/stdlib/src/collections/dict.mojo#L700

To be honest I am not sure if probing algorithm will actually finish at all, as after 13 steps where it “burns through the hash value”, it will probe in (slot * 5 + 1) % n. I guess it would be possible to write a mathematical proof, but it’s a bit over my head.

So in my opinion there are two options, implement probing with “while True”, or maybe even just “for _ in range(self._reserved + 13)”, but honestly I would rather implement linear probing instead. 

One more thing, while I looked at the code, I this branch https://github.com/modularml/mojo/blob/457378c3c5bafa287d3fc6be3a913e71e2233231/stdlib/src/collections/dict.mojo#L704 is also broken.

Long story short, I can create a PR, question is keep the probing as is and fix iterations, or switch to linear probing?

**Comment by bethebunny** [comment]
Hi @mzaks, thanks so much for tracking this down! There's definitely a couple bugs going on here, this is really awesome work :D (Also, [nice issue number](https://en.wikipedia.org/wiki/Taxicab_number)!)

Bug 1: As you correctly point out, because we're including higher order bits of the hash to perturb the probing sequence, the probing sequence isn't guaranteed to terminate in only `self._reserved` steps.
Bug 2: Because we "compact" the entries list and keep the load factor below 1, there will always be at least 1 empty slot in the dictionary when we run `_find_index`. This means assuming bug 1 is fixed we should deterministically return by the end of the probing sequence, and so the `insert_slot`, `insert_index`, and the text of the `elif index == Self.REMOVED` block are all never used.
Bug 3: In the specific case where a) we don't find an empty slot before "using up" the hash in `perturb` and b) the hash value is negative, `perturb` becomes `-1`, and the perturb equation becomes `5*i` rather than `5*i + 1`, which is _not_ guaranteed to visit every index. This case leads to an infinite loop.

I understand the intuition to avoid this complexity with something like linear probing, but linear probing has some important practical downsides that we want to avoid. It depends much more directly on the implementation of the hash function. For instance, suppose I have a type that's best represented as bit flags, represented as powers of 2, and choose to hash them to themselves. Since `Dict` always has a power of 2 size, larger flags will always have their initial probe index as 0, and this type will degrade to an O(n^2) worst case performance with linear probing, whereas a probe sequence that factors in the higher values of the hash will still achieve average-case linear time.

Per your comment it'd be nice to have a sketch of a proof in the code comments that iterations of `5*i + 1 % len(dict)` will probe every index. I can follow up with one a bit later (unless someone else wants to chime in!), it's been a while since my algebra classes but I think this has something to do with the ideals of polynomial rings.

My recommendation is
1) Update the logic of the perturb value shift to be a logical shift. I want to dig into this a bit to understand why we're not using a logical shift for `Int.__rshift__` and whether that makes sense (Python also does this so there's plausibly a good reason). It might make the most sense to make these values explicitly unsigned, or something else.
2) Update the logic of `_find_index` to `while True:` with some good comments, and simplify the remainder of the function to remove the unreachable code.

I've tested a hacky solution of this locally and verified it fixes all of these bugs, but if you want to submit a cleaner version with some tests that explicitly check your issue that'd be huge!

**Comment by mzaks** [comment]
@bethebunny I created a PR with a fix, which includes a hash function independent unit test. 

I fixed Bug 3 by insisting on `perturb` being a UInt64.

I fixed Bug 1 by switching to `while True:` it would also be possible to switch to `for _ in range(self._reserved + 13):` but IMHO it complicates thing as would need to ad an `abort("should never happen")` statement after the for loop.

Regarding Bug 2, the branch with `elif index == Self.REMOVED` is actually needed for the `pop` method. I simplified it though.

Generally regarding the probing algorithm, I think when this proposal of mine https://github.com/modularml/mojo/pull/2250 is approved and addressed, we can revisit this discussion. As with a good hash function the low bit being 0 should not happen. Linear probing has it's benefits in regards to memory access pattern and we could even experiment with SIMD, although I guess with random probing we could also first compute next slots and then gather the indices, but IMHO linear should be simpler and probably faster.

**Comment by bethebunny** [comment]
@mzaks I finally tracked down why this probe sequence is guaranteed to visit all locations! The magic words are "[linear congruential generator](https://en.wikipedia.org/wiki/Linear_congruential_generator#m_a_power_of_2,_c_%E2%89%A0_0)".
